crosscluster/logical: set UDF function name from SQL input

stevendanna opened the pull request:
First commit is #127024

See individual commits.

stevendanna committed:
crosscluster/logical: set UDF function name from SQL input

This wires up the function from the SQL syntax to the processor.

Epic: none
Release note: None

stevendanna committed:
crosscluster/logical: remove upsert_specified applier action

Most practically, this feature is what required the extra type which
is a bit annoying. Further, as shown in one of the updated tests, the
user can insert to the table manually if they want.

Release note: None
Epic: none

dt commented on pkg/ccl/crosscluster/logical/lww_row_processor.go:
```diff
@@ -192,17 +190,53 @@ func (q *queryBuffer) ApplierQueryForRow(row cdcevent.Row) (queryBuilder, error)
+			return errors.Errorf("cross-database replication job is not allowed")
+		}
+		if hasFunction != (conf.dstFnName != "") {
+			return errors.Errorf("function must be specified for all tables")
```
huh... why?

stevendanna replied:
Actually though, the code to do this is thinner than I thought. Will push it to this PR.

dt commented on pkg/ccl/crosscluster/logical/lww_row_processor.go:
```diff
@@ -192,17 +190,53 @@ func (q *queryBuffer) ApplierQueryForRow(row cdcevent.Row) (queryBuilder, error)
+			first = false
+		}
+		if dbName != conf.dstDBName {
+			return errors.Errorf("cross-database replication job is not allowed")
```
this seems very late to be validating this? shouldn't we be doing this before we create the job?

stevendanna replied:
I moved this. There is another bit of validation we should probably move early as well about the function living in the same database as the tables.

stevendanna committed:
crosscluster/logical: move and relax validation logic

Epic: none
Release note: None

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127428)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @azhu-crl, @dt, and @stevendanna)

___
*[`pkg/ccl/crosscluster/logical/logical_replication_job.go` line 379 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/127428#-O25hlB1BDwvL6N2zmwv:-O25hlB1BDwvL6N2zmww:b-evcvq7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d366ab935525f1843b33dc609e5b12e134f215e/pkg/ccl/crosscluster/logical/logical_replication_job.go#L379)):*
> ```Go
> 				// processors. This makes it a fragile to
> 				// renames.
> 				fnName = fmt.Sprintf("%s.%s",
> ```

we could pass it by OID instead. in case you weren't aware, we already have this syntax:

```
root@localhost:26257/defaultdb> CREATE FUNCTION add(a INT, b INT) RETURNS INT LANGUAGE SQL AS $$ SELECT a + b; $$;
CREATE FUNCTION

Time: 28ms total (execution 20ms / network 8ms)

root@localhost:26257/defaultdb> select 'add'::regproc::oid;
   oid
----------
  100105
(1 row)

Time: 2ms total (execution 2ms / network 1ms)

root@localhost:26257/defaultdb> SELECT [FUNCTION 100105](1, 2);
  add
-------
    3
(1 row)
```


<!-- Sent from Reviewable.io -->

dt commented:
Oh nice, I like that via-OID call that sidesteps all resolution messiness.

stevendanna commented:
Oh cool, it works across databases as well. Very useful!

stevendanna committed:
crosscluster/logical: call function by OID

This cleans up a lot of the work we were doing just to get the
database name so that we could call the function by name.

Epic: none
Release note: None

stevendanna commented:
Well that cleans things up rather nicely.

Jeremyyang920 commented on pkg/sql/execinfrapb/processors_bulk_io.proto:
```diff
@@ -485,6 +485,9 @@ message TableReplicationMetadata {
   optional string destination_parent_database_name = 2 [(gogoproto.nullable) = false];
   optional string destination_parent_schema_name = 3 [(gogoproto.nullable) = false];
   optional string destination_table_name = 4 [(gogoproto.nullable) = false];
+  // DestinationFunctionOid, if non-zoro, is the name of the
```
```suggestion
  // DestinationFunctionOid, if non-zero
```
Should name be OID here in this comment since that's what it is stored as?

stevendanna replied:
I kept going back and forth. The type is `oid.Oid`. But I suppose all of our functions use OID.

stevendanna commented:
bors r=dt

msbutler commented:
@stevendanna i believe the backport for this needs to merge after https://github.com/cockroachdb/cockroach/pull/127538
